File: Assignment-4/1/coin_change.py
from sys import maxsize as MAX

def change(money, coins):
    min_no_of_coins = [0] * (money + 1)
    #min_no_of_coins[0] = 0
    no_of_coins = 0
    for i in range(1 , money + 1):
        min_no_of_coins[i] = MAX
        for j in range(0 , 3):
            if(i >= coins[j]):
                num_of_coins = min_no_of_coins[i - coins[j]] + 1
            if(num_of_coins < min_no_of_coins[i]):
                min_no_of_coins[i] = num_of_coins
    return min_no_of_coins[money]

def main():
    money = int(input())
    coins = [1, 3, 4]
    print(change(money, coins))

if __name__ == "__main__":
    main()

File: Assignment-4/1/coin_change.cpp
#include<iostream>
#include<vector>
#include<limits>

using namespace std;

#define MAX numeric_limits<int>::max();

int coins[] = {1, 3, 4};

int get_change(int money)
{
    vector<int> minimum_no_coins(money + 1);
    int num_of_coins = MAX;
    minimum_no_coins[0] = 0;

    for(int i = 1; i <= money; i++)
    {
        minimum_no_coins[i] = MAX;
        for(int j = 0; j < 3; j++)
        {
            if(i >= coins[j])
            {
                num_of_coins = minimum_no_coins[i - coins[j]] + 1;
            }
            if(num_of_coins < minimum_no_coins[i])
            {
                minimum_no_coins[i] = num_of_coins;
            }
        }
        num_of_coins = 0;
    }
    return minimum_no_coins[money];
}

int main()
{
    int money = 0;

    cin>>money;

    cout<<get_change(money)<<endl;
    return 0;
}